lib/lte_link_control/lte_lc_helpers.c

```diff
@@ -30,6 +30,21 @@ struct event_handler {
 
 static sys_slist_t handler_list;
 
+struct lte_lc_event_fifo_item {
+	void *fifo_reserved;
+	struct lte_lc_evt evt;
+};
+
+static void lte_lc_event_work_fn(struct k_work *work);
+
+#define LTE_LC_EVENT_HEAP_SIZE 512 /* TODO: Make size Kconfig option */
+
+static K_FIFO_DEFINE(lte_lc_event_fifo);
+static K_HEAP_DEFINE(lte_lc_event_heap, LTE_LC_EVENT_HEAP_SIZE);
```

Comment:
Instead of using separate heap for storing the events, how about using a ring buffer? The Zephyr ring buffer seems to also have a data item mode. Or maybe a message queue, which I believe internally uses the ring buffer implementation.

Reply:
I think FIFO is targeted more for arbitrary size data. The events are fixed size, so there are probably simpler solutions available, which don't need to dynamically allocate memory.

Reply:
Yep, this was just copied from AT monitor quickly and we can change the structure and where it's allocated.

Reply:
> I think FIFO is targeted more for arbitrary size data. The events are fixed size, so there are probably simpler solutions available, which don't need to dynamically allocate memory.

When you don't dynamically allocate the memory, you need to specify the maximum size.
I think the best would be if we could guarantee certain small size and then use extra memory if available with k_malloc. Same discussion should apply to `at_monitor`.

Reply:
> When you don't dynamically allocate the memory, you need to specify the maximum size.
> I think the best would be if we could guarantee certain small size and then use extra memory if available with k_malloc. Same discussion should apply to at_monitor.

It's not different from the original implementation, where the `LTE_LC_EVENT_HEAP_SIZE` is the limiting factor. In practice that defines the maximum number of events which can be stored. Even though the allocation is dynamic, the pool from which the allocation is done is static.